File: java/Glacier2/session/server/src/main/java/com/example/glacier2/session/server/SharedPokeBox.java
// Copyright (c) ZeroC, Inc.

package com.example.glacier2.session.server;

import com.example.catchthemall.PokeBox;
import com.zeroc.Ice.Current;
import com.zeroc.Ice.DispatchException;
import com.zeroc.Ice.ReplyStatus;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * {@code SharedPokeBox} is an Ice servant that implements Slice interface 'PokeBox'. The same shared servant
 * implements all PokeBox objects; this is doable because all the state is stored in the {@link PokeStore}.
 */
class SharedPokeBox implements PokeBox {
    private final PokeStore _pokeStore;
    private final UserIdResolver _userIdResolver;

    /**
     * Constructs a {@code SharedPokeBox} servant.
     *
     * @param pokeStore the Poke-store
     * @param userIdResolver the user ID resolver
     */
    SharedPokeBox(PokeStore pokeStore, UserIdResolver userIdResolver) {
        _pokeStore = pokeStore;
        _userIdResolver = userIdResolver;
    }

    // Retrieve the Pokémon collection for the user associated with the current session.
    @Override
    public List<String> getInventory(Current current) {
        return _pokeStore.retrieveCollection(getUserId(current));
    }

    // Add new Pokémon to the Pokémon collection for the user associated with the current session.
    @Override
    public void caught(List<String> pokemon, Current current) {
        String userId = getUserId(current);

        List<String> newPokemon = new ArrayList<>(pokemon);
        List<String> savedPokemon = _pokeStore.retrieveCollection(userId);
        if (savedPokemon != null) {
            newPokemon.addAll(savedPokemon);
        }

        // Sort the collection to make it easier to see the Pokémon that have been caught.
        Collections.sort(newPokemon);
        _pokeStore.saveCollection(userId, newPokemon);
    }

    @Override
    public void releaseAll(Current current) {
        _pokeStore.saveCollection(getUserId(current), Collections.emptyList());
    }

    /**
     * Retrieves the user ID associated with the current session.
     *
     * @param current information about the incoming request being dispatched
     * @return the user ID associated with the current session.
     */
    private String getUserId(Current current) {
        String userId = _userIdResolver.getUserId(current.id.name);
        if (userId == null) {
            throw new DispatchException(ReplyStatus.Unauthorized.value(), "Invalid session token");
        }
        return userId;
    }
}

File: swift/Glacier2/Session/Sources/Server/SharedPokeBox.swift
// Copyright (c) ZeroC, Inc.

import Ice

/// SharedPokeBox is an Ice servant that implements Slice interface PokeBox. The same shared servant
/// implements all PokeBox objects; this is doable because all the state is stored in the PokeStore.
struct SharedPokeBox: PokeBox {
    private let pokeStore: PokeStore
    private let userIdResolver: UserIdResolver

    /// Constructs a SharedPokeBox servant.
    /// - Parameters:
    ///   - pokeStore: The Pokémon store.
    ///   - userIdResolver: The user ID resolver.
    init(pokeStore: PokeStore, userIdResolver: UserIdResolver) {
        self.pokeStore = pokeStore
        self.userIdResolver = userIdResolver
    }

    func getInventory(current: Ice.Current) async throws -> [String] {
        try await pokeStore.retrieveCollection(userId: getUserId(current: current))
    }

    func caught(pokemon: [String], current: Ice.Current) async throws {
        let userId = try await getUserId(current: current)
        var savedPokemon = await pokeStore.retrieveCollection(userId: userId)

        savedPokemon.append(contentsOf: pokemon)

        // Sort the collection to make it easier to see the Pokémon that have been caught.
        savedPokemon.sort()
        await pokeStore.saveCollection(userId: userId, pokemon: savedPokemon)
    }

    func releaseAll(current: Ice.Current) async throws {
        try await pokeStore.saveCollection(userId: getUserId(current: current), pokemon: [])
    }

    /// Retrieves the user ID associated with the current session.
    /// - Parameter current: Information about the incoming request being dispatched.
    /// - Returns: The user ID associated with the current session.
    /// - Throws: DispatchException if the session token is invalid.
    private func getUserId(current: Ice.Current) async throws -> String {
        guard let userId = await userIdResolver.getUserId(current.id.name) else {
            throw Ice.DispatchException(replyStatus: ReplyStatus.unauthorized.rawValue)
        }
        return userId
    }
}
